readers/multishard.cc

```diff
@@ -609,7 +609,9 @@ future<> evictable_reader_v2::fill_buffer() {
         auto* next_mf = co_await _reader->peek();
 
         // First make sure we've made progress w.r.t. _next_position_in_partition.
-        while (next_mf && _tri_cmp(_next_position_in_partition, buffer().back().position()) <= 0) {
+        // This loop becomes inifinite when next pos is a partition start.
+        // In that case progress is guranteed anyway, so skip this loop entirely.
+        while (!_next_position_in_partition.is_partition_start() && next_mf && _tri_cmp(_next_position_in_partition, buffer().back().position()) <= 0) {
```

Comment:
I don't understand the comment. Why is it guaranteed we make progress on partition_start?

Reply:
I see. Though I don't see why the current code is broken. You read the partition start and emit it, Then you read the partition end, which should have its position_in_partition greater than the partition start, which should stop the loop.

Reply:
It's absolutely broken to compare position_in_partition without noticing that we changed partitions, as position_in_partition can't be compared across partitions. So I agree with the fix, just wondering why partition_end didn't save us.

Reply:
Reading an entire partition into memory is an OOM sentence if the partition is large.

Reply:
Also note the other condition in the loop:
```
_tri_cmp(_next_position_in_partition, buffer().back().position()) <= 0
```

This will be true for all possible partition fragment kinds, as partition-start has the smallest possible position.

Reply:
But why would we read the entire partition? As soon as we read a clustering row or a partition_end the original condition would be satisfied.

In fact I don't see how the original condition can fail unless we have tons of range tombstones with the same position. If you read a clustering row, you make progress. If you read a partition end, you make progress. If you read a partition_start, the original condition thinks you didn't make progress, but the very next fragment will satisfy the condition.

Reply:
See the comment above the enclosing `if`:
```c++
    // Ensure that each buffer represents forward progress. Only a concern when
    // the last fragment in the buffer is range tombstone change. In this case
    // ensure that:
    // * buffer().back().position() > _next_position_in_partition;
    // * _reader.peek()->position() > buffer().back().position();
```
This loops ensures the former condition. We keep reading new fragments while `_next_position_in_partition <= buffer().back().position()`. This condition is always true if `_next_position_in_partition` is `partition_start`.

Reply:
But doesn't buffer.back() change during the loop? Ah, I guess it doesn't.

Reply:
I don't get it. The code is `while (_next_position_in_partition <= back) advance_back...`

But it this condition is supposed to check for progress against `_next_position_in_partition`, shouldn't it be `while (back <= _next_position_in_partition) advance_back...`?

Reply:
Yup. If you modify the test below to call `fill_buffer()` a second time, this time it will read till the end of partition instead of reading another small batch of fragments - because the comparison is done in the wrong direction.